analyzer/src/main/kotlin/managers/Yarn2.kt

```diff
@@ -126,21 +192,530 @@ class Yarn2(
+                "Locator '$locator' cannot be parsed."
+            }
+            val locatorRawName = locatorMatcher.groupValues[1]
+            val locatorType = locatorMatcher.groupValues[2]
+            val locatorVersion = locatorMatcher.groupValues[3]
+
+            val (locatorNamespace, locatorName) = Npm.splitNamespaceAndName(locatorRawName)
+            val version = cleanYarn2VersionString(locatorVersion)
+
+            val identifierType = if ("workspace" in locatorType) "Yarn2" else "NPM"
+            when {
+                // Prevent @virtual dependencies because they are internal to Yarn.
+                // See https://yarnpkg.com/features/protocols
+                "virtual" in locatorType -> null
+
+                else -> {
+                    try {
+                        Identifier(identifierType, locatorNamespace, locatorName, version)
+                    } catch (x: IllegalArgumentException) {
+                        x.showStackTrace()
+                        log.error { "Cannot build identifier for dependency '$locator.'" }
+                        null
+                    }
+                }
+            }
+        }.toList()
+
+    /**
+     * Parse the [definitionFile] (package.json) to find the scope of a dependency. Unfortunately, `yarn info -A -R`
+     * does not deliver this information.
+     * Return the dependencies present in the file mapped to their scope.
+     * See also https://classic.yarnpkg.com/en/docs/dependency-types (documentation for Yarn 1).
+     */
+    private fun listDependenciesByType(definitionFile: File): Map<String, DependencyType> {
+        val json = jsonMapper.readTree(definitionFile)
+        val result = mutableMapOf<String, DependencyType>()
+        DependencyType.values().forEach { dependencyType ->
+            json[dependencyType.type]?.fieldNames()?.asSequence()?.forEach {
+                result += it to dependencyType
+            }
+        }
+        return result
+    }
+
+    /**
+     * Clean the [rawVersion] string contained in a Yarn 2+ locator to have it compatible with NPM/Semver.
+     */
+    private fun cleanYarn2VersionString(rawVersion: String): String {
+        // 'Patch' locators are complex expressions such as
+        // resolve@npm%3A2.0.0-next.3#~builtin<compat/resolve>%3A%3Aversion=2.0.0-next.3&hash=07638b
+        // Therefore, the version has to be extracted (here '2.0.0-next.3).
+        var result = rawVersion.substringAfter("version=")
+            .substringBefore("&")
+
+        // Remove the archive URLs that can be present due to private registries.
+        // See https://github.com/yarnpkg/berry/issues/2192.
+        result = result.substringBefore("::__archiveUrl")
+
+        // Rewrite some dependencies to make them compatible with Identifier.
+        // E.g. typescript@patch:typescript@npm%3A4.0.2#~builtin<compat/typescript>::version=4.0.2&hash=ddd1e8
+        return result.replace(":", "%3A")
```

Comment:
Well I don't want to encode the full string to avoid incompatibility between Yarn2+ locator and NPM version strings.
I am only doing the minimal change which is getting rid of the `:` for the `Identifier` class.